File: apps/web/src/lib/auth/password-policy.ts
/**
 * PathForge — Shared Password Policy
 * ====================================
 * Single source of truth for password complexity validation on the
 * client. Mirrors ``apps/api/app/core/password_policy.py`` on the
 * backend — keep the two in sync whenever the ruleset changes.
 *
 * OWASP-aligned requirements (ASVS V2.1 — Password Security):
 * - Minimum 8 characters
 * - At least one uppercase letter
 * - At least one digit
 * - At least one special character from a wide, printable set
 *
 * Sprint 39 audit finding: the register and reset-password pages had
 * subtly different regexes (escape drift), so a password accepted by
 * one form could be rejected by the other. Extracting this module
 * fixes that drift; every form imports the same validator.
 */

// Keep this pattern in sync with
// ``apps/api/app/core/password_policy.py::PASSWORD_SPECIAL_CHARS_PATTERN``.
//
// Character class: !@#$%^&*(),.?":{}|<>-_=+[]\/'`~;
// Excludes whitespace and control characters; includes the full set of
// printable ASCII punctuation common to US-ASCII keyboards.
export const PASSWORD_SPECIAL_CHARS_REGEX = /[!@#$%^&*(),.?":{}|<>\-_=+[\]\\/'`~;]/;
export const PASSWORD_UPPERCASE_REGEX = /[A-Z]/;
export const PASSWORD_DIGIT_REGEX = /\d/;

export const PASSWORD_MIN_LENGTH = 8;
export const PASSWORD_MAX_LENGTH = 128;

/**
 * Validate that a password satisfies PathForge's complexity rules.
 *
 * Returns ``null`` on success, or a user-facing error message on
 * failure. The message format is deliberately identical to the
 * backend's validator so that errors look the same regardless of
 * which layer catches them.
 */
export function validatePasswordComplexity(value: string): string | null {
  if (value.length < PASSWORD_MIN_LENGTH) {
    return `Password must be at least ${PASSWORD_MIN_LENGTH} characters`;
  }
  if (value.length > PASSWORD_MAX_LENGTH) {
    return `Password must be at most ${PASSWORD_MAX_LENGTH} characters`;
  }

  const missing: string[] = [];
  if (!PASSWORD_UPPERCASE_REGEX.test(value)) missing.push("one uppercase letter");
  if (!PASSWORD_DIGIT_REGEX.test(value)) missing.push("one digit");
  if (!PASSWORD_SPECIAL_CHARS_REGEX.test(value)) missing.push("one special character");

  if (missing.length > 0) {
    return `Password must contain at least ${missing.join(", ")}`;
  }

  return null;
}

File: apps/api/app/core/password_policy.py
"""
PathForge — Shared Password Policy
====================================
Single source of truth for password complexity validation.

Used by both the register schema and the password-reset schema so that
the same rules apply everywhere. The frontend mirrors this regex in
``apps/web/src/lib/auth/password-policy.ts`` — keep the two in sync
whenever the ruleset changes.

OWASP-aligned requirements (ASVS V2.1 — Password Security):
- Minimum 8 characters (enforced at Field level via min_length)
- At least one uppercase letter
- At least one digit
- At least one special character from a wide, printable set

Why extract:
- Sprint 39 audit finding: frontend and backend regexes had drifted
  (subtle escaping differences). Users saw "valid on frontend, 422 from
  backend" errors. Extracting this constant fixes the drift.
"""

from __future__ import annotations

import re

# ── Special character class ────────────────────────────────────────────
# Keep this pattern in sync with
# ``apps/web/src/lib/auth/password-policy.ts``.
#
# Character class: !@#$%^&*(),.?":{}|<>-_=+[]\/'`~;
# Rationale: mirrors the "punctuation" set common to US-ASCII keyboards;
# excludes whitespace and control characters. `-` is placed after an
# escape (`\-`) for portability across regex dialects.
PASSWORD_SPECIAL_CHARS_PATTERN: str = r"[!@#$%^&*(),.?\":{}|<>\-_=+\[\]\\/'`~;]"

_UPPERCASE_PATTERN: re.Pattern[str] = re.compile(r"[A-Z]")
_DIGIT_PATTERN: re.Pattern[str] = re.compile(r"\d")
_SPECIAL_PATTERN: re.Pattern[str] = re.compile(PASSWORD_SPECIAL_CHARS_PATTERN)


def validate_password_complexity(value: str) -> str:
    """Raise ``ValueError`` if ``value`` does not satisfy complexity rules.

    Returns the input unchanged on success so callers can use this as a
    Pydantic ``@field_validator`` body or a pre-save hook.

    The message format matches the frontend validator verbatim so that
    error messages are identical regardless of which layer catches them.
    """
    missing: list[str] = []
    if not _UPPERCASE_PATTERN.search(value):
        missing.append("one uppercase letter")
    if not _DIGIT_PATTERN.search(value):
        missing.append("one digit")
    if not _SPECIAL_PATTERN.search(value):
        missing.append("one special character")

    if missing:
        raise ValueError(
            f"Password must contain at least {', '.join(missing)}"
        )
    return value
